preview/FS-1137-prefer-extensions-method-over-property-when-arguments-provided.md

```diff
@@ -0,0 +1,115 @@
+
+## Implementation details
+
+When there is a "delayed application" expression adjacent to the identifier which would normally resolve to a property, the resolution considers extension methods / type extensions that are in the name resolution environment.
+
+## Documentation
+
+Documentation around extension methods and type extension may be updated to reflect what this RFC enables.
+
+## Tooling
+
+Editors have to decide how they want to list the members, the `DeclarationListInfo` type of FCS is being updated to not merge extension methods that may bear the same name as properties anymore in order to make the support look the same as what C# editor in VS does (property and extension methods having the same name are listed as separate intellisense entries).
+
+The implementation detail for this change carries a slight risk of having separate entries for unexpected members (in presence of methods defined in type extension or extension method) while before the feature, only the first entry by name would show.
+
+Non-regression tests can be added and the implementation detail adjusted, in case such an issue surfaces.
+
+Such an issue isn't expected to be disastrous to the user experience, while not having the separate entries make the feature much less discoverable.
+
+# Drawbacks
+
+## Properties that cannot be shadowed
+
+Due to the idiom of indexed property not being possible to express in C#, and the ambiguity with delayed application of expressions in case of properties having indexers, or being of a function type, there is no plan to support resolution of extension methods / type extensions whenever those case apply. This introduce a subtle inconsistency among properties that can be defined on F# types.
+
+## The features renders obsolete the lack of support for intrinsic methods of same name as a property
+
+In terms of language design, beside subtleties in overload resolution pertaining to methods, there is little reasons to distinguish intrinsic methods versus extension methods or methods defined in type extensions. Yet, C#, VB.NET and F# still preclude for methods to be defined with the same name as a property, despite:
+* they can't bear the same name in the IL (properties have a `get_` or `set_` prefix addorned to the IL methods implementation of getter and setter)
+* they can be defined as extension methods in C# & VB.NET, and in F# with this feature, as method in type extension and extension methods
+
+It may make sense to lift this restriction in case C# moves in this direction.
+
+In meantime, attempting to define such intrinsic methods still results in 
+
+>  error FS0434: The property '{property name}' has the same name as a method in type '{type name}'.
+
+## FCS breaking change in items returned `DeclarationListInfo`
```

Comment:
Do we know of any cases if this can be breaking change in runtime? e.g. if property exposes function or delegate?

```fsharp
open System.Runtime.CompilerServices
type TypeWithProperty () =
   member _.X
     with get () = fun (x:int) -> printfn $"Property: {x}"; x
[<Extension>]
type TypeWithPropertyExt =
     [<Extension>]
     static member X (t: TypeWithProperty, i: int) = printfn $"Method: {i}"; i
let t = TypeWithProperty()
t.X(1) // this will "call" property
```

Reply:
I presume it's not a breaking change in this sense, could you please make an explicit comment, that it's not?

Reply:
It still calls the property due to check on property type being `TType_fun` to ignore looking up extension methods.

https://github.com/dotnet/fsharp/blob/6a9791f6cc5f6e59ca729917e991b827d00541dc/src/Compiler/Checking/NameResolution.fs#L2644-L2646

I tried declaring events, they don't go to the same "Property" branch, and regarding properties that would be of a delegate type, in F# they are always invoked with `.Invoke` so it shortcuts as having no "delayed items" for which type checking works the same as before.

Based on how it is specified (things that can't be shadowed) and how it is implemented, I'm overall confident that it won't be a breaking change to existing code.

The edge case that would be breaking I thought about is indexed properties overloads for a non indexed property but F# doesn't allow to specify this:
```fsharp
type C() =
    member _.Y = 1
    member _.Y with get (v,v2) = printfn $"Y({v},{v2})"
```

> error FS0436: The property 'Y' has the same name as another property in type 'C', but one takes indexer arguments and the other does not. You may be missing an indexer argument to one of your properties.